src/ucp/api/ucp.h

```diff
@@ -3135,6 +3108,159 @@ ucs_status_ptr_t ucp_worker_flush_nb(ucp_worker_h worker, unsigned flags,
+                                               ) ;
+
+/*
+ * Structure for communicatiob between the active message
+ * callback and the ucp library. The ucp library sets up 'iovec_max_length'
+ * to indicate the length of the iovec in this structure; the remaining
+ * fields are set by the callback to indicate what should happen
+ * with the remaining data from the client. In the initial implementation,
+ * iovec_max_length is always 1.
+ */
+typedef struct ucp_am_recv {
+    /* Function to be driven when data transfer is complete */
+    ucp_am_local_function_t local_fn;
+    /* Argument to be passed to local_fn */
+    void                   *cookie;
+    /* Function to be driven when each fragment of data transfer is complete
+     * In the initial implementation, all data is transferred in one fragment, so
+     * the data_fn si driven once just before the local_fn
+     */
+    ucp_am_data_function_t  data_fn;
+    /* Argument to be passed to data_fn */
+    void                   *data_cookie ;
+    /* Sise of iovec allocated by the ucp library */
+    size_t                  iovec_max_length;
+    /* Size of iovec filled in by the callback */
+    size_t                  iovec_length;
+    /* iovec indicating where the data from the client is to be placed */
+    ucp_dt_iov_t            iovec[1];
+  } ucp_am_recv_t;
+
+  /**
+   * @ingroup UCP_ENDPOINT
+   * @brief Callback to process incoming Active Message.
+   *
+   * When the callback is called, @a flags indicates how @a data should be handled.
+   *
+   * @param [in]  arg      User-defined argument.
+   * @param [in]  data     Points to the received data header.
+   * @param [in]  length   Length of data header.
+   * @param [in]  reply_ep If the Active Message is sent with the
+   *                       UCP_AM_SEND_REPLY flag, the sending ep
+   *                       will be passed in. If not, NULL will be passed.
+   * @param [in]  flags    If this flag is set to UCP_CB_PARAM_FLAG_DATA,
+   *                       the callback can return UCS_INPROGRESS and
+   *                       data will persist after the callback returns.
+   * @param [in]  remaining_length Length of data which has not yet been
+   *                       transferred
+   * @param [out] recv     Struture passed to the ucp library to control
+   *                       placement of the remaining data
+   *
+   * @return UCS_OK        @a data will not persist after the callback returns.
+   *
+   * @return UCS_INPROGRESS Can only be returned if flags is set to
+   *                        UCP_CB_PARAM_FLAG_DATA. If UCP_INPROGRESS
+   *                        is returned, data will persist after the
+   *                        callback has returned. To free the memory,
+   *                        a pointer to the data must be passed into
+   *                        @ref ucp_am_data_release.
+   *
+   * @return UCS_ERR_NO_MEMORY Callback was unable to allocate memory for
```

Comment:
And how this will be handled ?

Reply:
Good question. I don't think ucx has meaningful handling for out-of-memory conditions; I would expect the applications at the endpoints will need to fail an assert.